Adds a Ginkgo-style unit test to `kn version` command

maximilien opened the pull request:
1. adds suites for each package where commands reside
2. adds version_test.go to completely unit test version
   as an example of what can be done

NOTE also that all current Go tests are still compatible
and runnable. So if we decide to start using this fashion
for UTs then we can gradually do so. I'll make some comments
about some advantages of Ginkgo/Gomega in issue #161 which
is the one we are using to track discussion.

maximilien commented:
/retest

maximilien commented:
`pull-knative-client-build-tests` are failing with:

```bash
I0613 23:47:27.285] -type Storage interface {
I0613 23:47:27.285] -	GetBucketFor(string) (*LeakyBucket, error)
I0613 23:47:27.285] -	SetBucketFor(string, LeakyBucket) error
...
```

Clearly not due to this PR since not related code and `pull-knative-client-integration-tests` and `pull-knative-client-unit-tests` jobs succeeded. 

Will retry tests tomorrow. Maybe changes are happening on the backend that need to stabilize.

rhuss commented:
/retest

rhuss commented:
@maximilien the test are failing because you have to update the deps first. Just run `hack/build.sh -u` and update the PR.

maximilien commented:
> Thanks a lot, looks like a great step in the right direction !
> As a ginko newcomer this all looks good :)For some minor issues see below.

👍 

> * I wouldn't call this _BDD style_ (as defined e.g. [here](https://en.wikipedia.org/wiki/Behavior-driven_development#Behavioral_specifications), where a BDD test contains a narrative with acceptance criteria (in plain text), translated to code. [Cucumber](https://cucumber.io/) would be such a framework. For me it looks more like 'just' unit testing framework combined with a fluent assertion library (like [assertj](https://joel-costigliola.github.io/assertj/)).

OK fair point. I was going by how the Ginkgo repo itself describes it.

> * It would be interesting to see how the real mocking/faking of the Knative API works in the context of this framework as this is imo the thing which is currently not so nicely solved in the current tests.

That's where [counterfeiter](https://github.com/maxbrunsfeld/counterfeiter) will come in. I will do all the plugins UTs and IT using Ginkgo/Gomega and Counterfeiter to test this. More next week.

Thanks for other feedback. Will address.

maximilien commented on pkg/kn/commands/commands_suite_test.go:
```diff
@@ -0,0 +1,47 @@
+var Output string
+var ReadFile, WriteFile *os.File
+
+var _ = BeforeSuite(func() {
```
I actually will change this. I think I can have it in a `JustBeforeEach` for the test itself. In the suite is useful for reuse. Assuming other tests in that package need it. We can always move if that presents itself.

maximilien replied:
Kept it in the suite but refactored. Should be easy to move as needed.

maximilien commented on pkg/kn/commands/commands_suite_test.go:
```diff
@@ -0,0 +1,47 @@
+	ReadFile, WriteFile, err = os.Pipe()
+	Expect(err).NotTo(HaveOccurred())
+	Stdout = WriteFile
+	os.Stdout = WriteFile
```
Indeed. Saw that. And an item to discuss. Why are not all output redirected in `KnParams.Output`. I will open an issue on this if it's not done. It should help also address #182 in a simpler fashion. Been wanting to have a central place for all communication to user. This is good time to do it.

maximilien commented on pkg/kn/commands/version_test.go:
```diff
@@ -0,0 +1,55 @@
+		It("Creates a VersionCommand", func() {
+			Expect(versionCmd).NotTo(BeNil())
+			Expect(versionCmd.Use).To(Equal("version"))
+			Expect(versionCmd.Short).To(Equal("Prints the client version"))
```
Yup. See [Gomega matchers](http://onsi.github.io/gomega/#provided-matchers). For contains see these: [ContainsSubstring](http://onsi.github.io/gomega/#containsubstringsubstr-string-args-interface) and [MatchRegexp](http://onsi.github.io/gomega/#matchregexpregexp-string-args-interface) and likely a few more. 

And if that's not enough you can create your own matcher or Golang directly.

maximilien commented on pkg/kn/commands/version_test.go:
```diff
@@ -0,0 +1,55 @@
+		It("Prints Version, Build date, GitRevision, and ServingVersion", func() {
+			err := versionCmd.RunE(nil, []string{})
+			Expect(err).NotTo(HaveOccurred())
+			Expect(ReadStdout()).To(Equal(versionOutput))
```
Yeah. Debated this a bit in my head. However, I figured this output has a specific format that we want to check for. I might change the `versionOutput` to be synthesized with the const. That way avoid dup and clear. Essentially `versionOuput` encapsulates the format of the output, more than anything.

maximilien commented on pkg/kn/commands/version_test.go:
```diff
@@ -0,0 +1,55 @@
+		versionCmd = NewVersionCommand(knParams)
+	})
+
+	Describe("NewVesionCommand", func() {
```
👍

maximilien commented:
OK see latest changes @rhuss. A bit overkill but addresses your points and should give more examples of how ginkgo/gomega helps with UT.

Also has updated deps (thx) and should be ready to merge on CI confirms.

cc: @sixolet @cppforlife

rhuss commented:
> OK fair point. I was going by how the Ginkgo repo itself describes it.

Well, it's a _bit_ like tests resulting in BDD style of coding as it has given-when-then sections. But they are quite distributed, and I think it does not promote writing the spec first (like in "Behaviour _driven_ development"), but more or less serves as a nice documentation for a test (which gets written before that spec). But this is confessedly a matter of perspective.

I just have written my first test for #134 to test the new kn client API in https://github.com/knative/client/pull/134/files#diff-e50a903452108905f85908deaf847f36 to get a feeling for the test library.

My impression its quite verbose and big and might be a bit confusing because of its functional style (you also can consider this just as a kind of DSL for your test-framework).

For me, it worked well, and I think the tests are also quite readable. 

However, some people report kind issues in parallelising tests because of all that deferred execution. This Reddit thread https://www.reddit.com/r/golang/comments/7a8zbs/why_is_ginkgo_so_popular/ might be a good reflection of various opinions on BDD. (Also intereting in this thread: @thockin mentions that Ginkgo is used genuinely in Kubernetes itself, but he seems [not so happy with this](https://www.reddit.com/r/golang/comments/7a8zbs/why_is_ginkgo_so_popular/dp89p8h/))

I also found this nice [list of golang test frameworks](https://bmuschko.com/blog/go-testing-frameworks/), which also mentions Ginkgo.

Said that until now I hadn't found anything fundamentally bad with Gingko (and I like indeed the fluent gomega assertion library), so I would start with Ginkgo if there are no other objections.

maximilien commented:
Nice analysis and thanks for the links. Will take a look and discuss with @onsi next time I see him.

I've never questioned much of Ginkgo/Gomega since I've sort of "grew up" with it since starting to program Golang for Cloud Foundry 5 years ago or so. 

The framework has gotten better and better and as I mentioned everything in CF (at least the Golang code parts) is tested with it. 

So my confidence level on the quality and usage and scalability of the framework is sky high.

> Said that until now I hadn't found anything fundamentally bad with Gingko (and I like indeed the fluent gomega assertion library), so I would start with Ginkgo if there are no other objections.

Nice. I think it will help with our tests alot. If there are places where it's not so nice or something better could be done, let's discuss. For now, just having a solid set of UTs and more importantly ITs are key.

Cheers 🍻

cppforlife commented:
sounds like this needs to be discussed at the WG. @maximilien can you add an agenda item?

maximilien commented:
> sounds like this needs to be discussed at the WG. @maximilien can you add an agenda item?

Yup, but not clear it makes sense for me this week.  Busy with stuff at KubeCon Shanghai. However, next week I can make time while in Beijing. Less scheduled parties 🎉😄

rhuss commented:
I did now a deep dive with Ginkgo and a more lightweight alternative based on [testing](https://golang.org/pkg/testing/) while rebasing https://github.com/knative/client/pull/134

I implemented the non-trivial tests for this refactoring in both Ginkgo and plain testing with some support by an assertion library (gotest.tools in this case, could have been also "testify").

You see both tests here:

* PR #214 Ginkgo + Gomega --> https://github.com/rhuss/client/blob/pr/add-schema-handling-ginko-test/pkg/serving/v1alpha1/client_test.go
* PR #134 testing + gotest.tools --> https://github.com/rhuss/client/blob/pr/add-schema-handling/pkg/serving/v1alpha1/client_test.go

I also read several blogs and found the one from @vdemeester really good and reasonably recent.

Let me summarize my observations and assessments:

### Ginkgo + Gomega

* Introducing these frameworks is quite intrusive as it adds around ~ 280 files (the PR with Gingko has 298 files changed)
* Ginkgo uses a new domain-specific language (DSL) for defining tests which for the casual reader is not easy to understand, and certain conventions should be obeyed. However, if you learned it is easy to understand and quite expressive. There are still ramp-up costs for new contributors.
* For our use cases for testing Ginkgo doesn't bring much new on the table which can be realized with other means. The assertions via Gomega are nice, but a bit verbose and can be as easy realized by other assertion libraries. 
* Grouping of tests is a nice feature of Ginkgo with its "Concepts"
* BDD kind of descriptions is not really of much value for us if we don't define "[Given-When-Then](https://martinfowler.com/bliki/GivenWhenThen.html)" user stories in advance. This is very uncommon for a community project, so I doubt that this will happen.
* Ginkgo has a nice reporting of test results, which are easy to understand.
* If not everything is written in Ginkgo we have a mixed reporting of the results. And also sometimes a bit confusing as the Ginkgo test report a success although another non-Gingko test fails.

### testing + gotest.tools (or any other assertion library)

* Reasonably lightweight (~ 40 files, the reference PR with gotest.tools has 59 files)
* Tests are still using `testing.T`, so it's very easy to onboard gophers.
* All test results look the same
* Grouping can be realized with subtests (see the examples)
* BDD kind of descriptions can be added via description to subtests (also see the example). They will show up when a test fails.
* No extra DSL is used. Its all go.
* Assertions are shorter and less verbose
* A gut feeling (not tested, but could do so if requested): Testing based tests are faster. And fast tests are important.
* Benefit of gotest.tools: It also contains icmd which is at the heart of the Docker CLI test suite and provides some help for testing CLI based commands.

### Summary

Now that I have implemented both, I clearly tend towards using plain testing with a lightweight assertion library. It's not that I say that Ginkgo has not its value (and we probably barely scratched the surface, there must be much more hidden in those 280 dependency files), but for an open community project introducing such a strongly opinionated framework with an own DSL adds some extra burden to contributors (remember, we require mandatory tests for contributions). Every contributor would need to learn Ginkgo first (definitely we should not drive a mixed-mode). Learning Ginkgo is not rocket science but takes a considerable amount of time. So we for sure will lose the 'casual contributor'. 

Whether the assertion library is gotest.tools or testify or something even simpler, I don't mind so much (although I like the flexibility of gotest.tools, its support for CLI testing and also its usage of e.g. google/go-cmp or support for async tests, and much more. It's a pick-and-choose approach though, not an all-in-one. See @vdemeester's [blog](https://vincent.demeester.fr/tags/#golang) for more details.)

But it would be good if we come to decission soonish. So +1 to put it on the next's WG meeting's topic.

maximilien commented:
We can definitely put this in this week's call or next one. Thanks for more analysis @rhuss

rhuss commented:
I added an agenda item to the WG doc with links to the relevant PRs and sample tests.

maximilien commented:
FYI... the failure in build is due to the lack of `gotest.tools` dependencies. Once @rhuss's #134 or any PR with that dependency is merged I will rebase and push back.

maximilien commented:
Actually, let me close this and submit a new one with `gotest.tools` it's cleaner.

maximilien commented:
Closing since we now have PR #223 that implements this in `gotest.tools` and also add tests for `completion` command.
